sql: add base functionality to create implicit index column partitioning

otan opened the pull request:
resolves #58727
resolves #58728 

# descpb: add num_implicit_columns field to PartitioningDescriptor

Release note: None

# sql: add experimental flag for using implicit columns for partitioning

Release note: None

# sql: add base functionality to create implicit index column partitioning

Introduced the ability to omit index columns when declaring partitions.
This is currently only available after setting the experimental flags
`experimental_enable_implicit_column_partitioning`.

None of this is hooked up to anything so I expect trying to use it will
crash a few things.

Also add unimplemented ALTER INDEX/TABLE on indexes/tables with implicit
partitioning -- that can be improved or implemented at a later date.

Release note: None

otan committed:
descpb: add num_implicit_columns field to PartitioningDescriptor

Release note: None

otan committed:
sql: add experimental flag for using implicit columns for partitioning

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/58521)
<!-- Reviewable:end -->

ajstorm reviewed (approved):
Good stuff!  Mostly minor stuff.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-:-MQoHykuDoMjFc9aVpg1:b-p5nnrq)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @miretskiy, and @otan)

___
*[pkg/sql/create_table.go, line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQoG7z6Bb2jIsH8auBh:-MQoG7z6Bb2jIsH8auBi:b-3ce8p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/create_table.go#L2414)):*
> ```Go
> 	seenImplicitColumnNames := map[string]struct{}{}
> 	var implicitColumnIDs []descpb.ColumnID
> 	var numImplicitColumns []string
> ```

Nit: Might just be me, but the use of `num` here is a bit confusing.  Isn't this just the list of implicit columns - `implicitColumns`?  Wouldn't `num` be `len(implicitColumns)`
___
*[pkg/sql/create_table.go, line 2417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQnxivZ0nAp7UtsCIuB:-MQnxiv_4cOj3CwlDzrM:bjmr21o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/create_table.go#L2417)):*
> ```Go
> 	var implicitColumnDirections []descpb.IndexDescriptor_Direction
> 	for _, field := range partBy.Fields {
> 		if string(field) == indexDesc.ColumnNames[0] {
> ```

Might be good to add a comment above this check.  Is the reason we're breaking out at this point because once we find the first index column we're done processing implicit columns
___
*[pkg/sql/catalog/descpb/structured.proto, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQoH2tU7Y3MVy1xpTFg:-MQoH2tU7Y3MVy1xpTFh:bbcrz87) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/catalog/descpb/structured.proto#L322)):*
> ```protobuf
>   // partition's columns. If NumColumns is 0, then there is no partitioning.
>   optional uint32 num_columns = 1 [(gogoproto.nullable) = false];
>   // NumImplicitColumns is how large of the prefix of columns in an index
> ```

Nit: Wording.  Would suggest:

"NumImplicitColumns specifies the number of columns that implicitly prefix a given index.  This is used for PARTITION ALL, which causes all indexes to be implicitly partitioned using the partitioning column(s) of the associated table. If NumImplicitColumns is 0, there are no implicit columns defined for the index."


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/58521)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajstorm and @miretskiy)

___
*[pkg/sql/create_table.go, line 2414 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQoG7z6Bb2jIsH8auBh:-MQoXzot460OzADccDvU:b-zgyilb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/create_table.go#L2414)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: Might just be me, but the use of `num` here is a bit confusing.  Isn't this just the list of implicit columns - `implicitColumns`?  Wouldn't `num` be `len(implicitColumns)`
</blockquote></details>

good catch, this was originally called a different field, i re-typed but did not re-name it.
___
*[pkg/sql/create_table.go, line 2417 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQnxivZ0nAp7UtsCIuB:-MQoYk8N3o5EDOR2FcyQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/create_table.go#L2417)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Might be good to add a comment above this check.  Is the reason we're breaking out at this point because once we find the first index column we're done processing implicit columns
</blockquote></details>

Done.
___
*[pkg/sql/catalog/descpb/structured.proto, line 322 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/58521#-MQoH2tU7Y3MVy1xpTFg:-MQoXQjqA1RVrqwylUxy:b-ydcyg9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8dd68b185095d85ddb579036e9c30f20159f3a2/pkg/sql/catalog/descpb/structured.proto#L322)):*
<details><summary><i>Previously, ajstorm wrote…</i></summary><blockquote>

Nit: Wording.  Would suggest:

"NumImplicitColumns specifies the number of columns that implicitly prefix a given index.  This is used for PARTITION ALL, which causes all indexes to be implicitly partitioned using the partitioning column(s) of the associated table. If NumImplicitColumns is 0, there are no implicit columns defined for the index."
</blockquote></details>

I've re-worded this as i realsied this is not just PARTITION ALL


<!-- Sent from Reviewable.io -->

otan committed:
sql: add base functionality to create implicit index column partitioning

Introduced the ability to omit index columns when declaring partitions.
This is currently only available after setting the experimental flags
`experimental_enable_implicit_column_partitioning`.

None of this is hooked up to anything so I expect trying to use it will
crash a few things.

Also add unimplemented ALTER INDEX/TABLE on indexes/tables with implicit
partitioning -- that can be improved or implemented at a later date.

Release note: None

otan commented:
bors r=ajstorm
